# 1. Exercise from course

## 1.1 If-statements with user-defined functions

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [6]:
# Import data set ords_prods_merge
path = r'/Users/minglunchen/Desktop/12-2023 Instacart Basket Analysis'

In [7]:
path

'/Users/minglunchen/Desktop/12-2023 Instacart Basket Analysis'

In [8]:
df_ords_prods_merge = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data', 'ords_prods_merge.pkl'))

In [12]:
# Create subset to avoid memory issues
df = df_ords_prods_merge[:1000000]

In [13]:
df.shape

(1000000, 16)

In [14]:
df.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  orders_interval  first_order  product_id  \
0                  8              NaN         True         196   
1                  7             15.0        False         196   
2                 12             21.0        False         196   
3                  7             29.0        False         196   
4                 15             28.0        False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [15]:
# Define a function
def price_label(row):
    
    if row['prices'] <= 5:
        return 'Low range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid range product'
    elif row['prices'] >15:
        return 'High range product'
    else: return np.nan

In [16]:
# Apply the function
df['price_range'] = df.apply(price_label, axis=1)

/var/folders/j4/lbgp6g5n3wj5p_xltrtb58hm0000gn/T/ipykernel_53516/902492192.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [17]:
df['price_range'].value_counts(dropna = False)

price_range
Mid range product    756450
Low range product    243550
Name: count, dtype: int64

In [18]:
df['prices'].max()

14.8

## 1.2 If-statements with the loc() Function

In [20]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High range product' 

In [24]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid range product' 

In [25]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low range prodcut'

In [26]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid range product     21890146
Low range prodcut     10126384
High range product      417682
Name: count, dtype: int64

## 1.3 If-statements with For-Loops

In [28]:
# An example of a simple loop
for x in range(30, 45):
    print("My age is %d" % (x))

My age is 30
My age is 31
My age is 32
My age is 33
My age is 34
My age is 35
My age is 36
My age is 37
My age is 38
My age is 39
My age is 40
My age is 41
My age is 42
My age is 43
My age is 44


In [29]:
# how busy is each day of the week: the frequency of orders in each day
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6209632
1    5665830
6    4500246
2    4217766
5    4209449
3    3844096
4    3787193
Name: count, dtype: int64

In [32]:
# create a column called 'result' to indicate how busy
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [33]:
df_ords_prods_merge['busiest_day'] = result

In [35]:
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22437387
Busiest day        6209632
Least busy         3787193
Name: count, dtype: int64

In [36]:
df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  orders_interval  first_order  product_id  \
0                  8              NaN         True         196   
1                  7             15.0        False         196   
2                 12             21.0        False         196   
3                  7             29.0        False         196   
4                 15             28.0        False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

  _merge    price_range_loc     busiest_day  
0   both  Mid range product  Regularly busy  
1   both  Mid range product  Regularly busy  
2   both  Mid range product  Regularly busy  
3   both  Mid range product      Least busy  
4   both  Mid range product      Least busy

# 2. Task

### Step 2: Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column using a suitable method.

In [37]:
# To identify how busy each day of the week is: the frequency of orders of each day
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6209632
1    5665830
6    4500246
2    4217766
5    4209449
3    3844096
4    3787193
Name: count, dtype: int64

In [53]:
# Create a new column
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
    if value == 0 or value == 1:
        result.append("Busiest days")
    elif value == 4 or value == 3:
        result.append("Slowest days")
    else:
        result.append("Regular days")

In [54]:
result

['Regular days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Regular days',
 'Regular days',
 'Regular days',
 'Slowest days',
 'Regular days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Slowest days',
 'Slowest days',
 'Regular days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regular days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regular days',
 'Slowest days

In [55]:
# Create a new column 'busiest_day_2' within the dataframe and set it equal to result
df_ords_prods_merge['busiest_day_2'] = result

In [57]:
# Check whether the new column is in the dataframe 
df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  orders_interval  first_order  product_id  \
0                  8              NaN         True         196   
1                  7             15.0        False         196   
2                 12             21.0        False         196   
3                  7             29.0        False         196   
4                 15             28.0        False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

  _merge    price_range_loc     busiest_day busiest_day_2  
0   both  Mid range product  Regularly busy  Regular days  
1   both  Mid range product  Regularly busy  Slowest days  
2   both  Mid range product  Regularly busy  Slowest days  
3   both  Mid range product      Least busy  Slowest days  
4   both  Mid range product      Least busy  Slowest days

### Step 3: Check the values of this new column for accuracy. Note any observations in markdown format.

In [58]:
df_ords_prods_merge['busiest_day_2'].value_counts(dropna = False)

busiest_day_2
Regular days    12927461
Busiest days    11875462
Slowest days     7631289
Name: count, dtype: int64

Notes: In the new column, 
'Busiest days' includes 0 and 1 -- the sum of their frequency is 6209632+5665830=11875462, which is identical with the frequency of 'Busiest days'.
Similarly, the frequency of 'Slowest days' is identical with the sum of 3 and 4; that of 'Regular days' is identical with the sum of 2, 5 and 6.

### Step 4: When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [59]:
# To identify how busy each hour of the day is: the frequency of orders of each day
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2764390
11    2738585
14    2691563
15    2664522
13    2663272
12    2620800
16    2537469
9     2456661
17    2089452
8     1719952
18    1637922
19    1259382
20     977017
7      891928
21     796362
22     634737
23     402612
6      290770
0      218942
1      115786
5       88057
2       69431
4       53283
3       51317
Name: count, dtype: int64

In [60]:
# Create a new column
result = []

for value in df_ords_prods_merge["order_hour_of_day"]:
    if value == 10 or value == 11 or value == 14 or value == 15 or value == 13 or value == 12 or value == 16 or value == 9:
        result.append("Most orders")
    elif value == 3 or value == 4 or value == 2 or value == 5 or value == 1 or value == 0 or value == 6 or value == 23:
        result.append("Fewest orders")
    else:
        result.append("Average orders")

In [61]:
# Create a new column 'busiest_period_of_day' within the dataframe and set it equal to result
df_ords_prods_merge['busiest_period_of_day'] = result

In [62]:
# Check whether the new column is in the dataframe 
df_ords_prods_merge.head()

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   

   order_hour_of_day  orders_interval  first_order  product_id  \
0                  8              NaN         True         196   
1                  7             15.0        False         196   
2                 12             21.0        False         196   
3                  7             29.0        False         196   
4                 15             28.0        False         196   

   add_to_cart_order  reordered product_name  aisle_id  department_id  prices  \
0                  1          0         Soda        77              7     9.0   
1                  1          1         Soda        77              7     9.0   
2                  1          1         Soda        77              7     9.0   
3                  1          1         Soda        77              7     9.0   
4                  1          1         Soda        77              7     9.0   

  _merge    price_range_loc     busiest_day busiest_day_2  \
0   both  Mid range product  Regularly busy  Regular days   
1   both  Mid range product  Regularly busy  Slowest days   
2   both  Mid range product  Regularly busy  Slowest days   
3   both  Mid range product      Least busy  Slowest days   
4   both  Mid range product      Least busy  Slowest days   

  busiest_period_of_day  
0        Average orders  
1        Average orders  
2           Most orders  
3        Average orders  
4           Most orders

### Step 5: Print the frequency for this new column.

In [63]:
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most orders       21137262
Average orders    10006752
Fewest orders      1290198
Name: count, dtype: int64

### Step 7: Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [64]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_derived.pkl'))